os: implement readdir for darwin and linux

dkegel-fastly opened the pull request:
Other operating systems are left as an exercise for the reader :-)

dkegel-fastly committed:
os: implement readdir for darwin and linux

readdir is disabled on linux for 386 and arm until syscall.seek is implemented there.

windows is hard, so leaving that for later.

File src/os/dir_other_go115.go can be deleted when we drop support for go 1.15.

Also adds TestReadNonDir, which was helpful while debugging.

dkegel-fastly commented:
The i386 problem this ran into in CI may be related to https://github.com/tinygo-org/tinygo/issues/1906

It can be reproduced with the commands

```
make
cd tests/testing/pass
GOARCH=386 ../../../build/tinygo test
```
in this branch, but not in dev.

dgryski reviewed (approved):
LGTM.  Large patch, but almost all of it seems to be from upstream Go. And it has tests, which pass.  👍

deadprogram commented:
Something not working quite right here: https://github.com/tinygo-org/tinygo/runs/4845686957?check_suite_focus=true#step:17:420

```
tinygo:ld.lld: error: undefined symbol: syscall.seek
```

dkegel-fastly commented:
Yes, see https://github.com/tinygo-org/tinygo/pull/2499#issuecomment-1013768071

The real fix may involves Ayke's fix for https://github.com/tinygo-org/tinygo/issues/1906

Until then, maybe I could figure out some clever way to exclude the architectures with the problem...?

dkegel-fastly commented:
Needs a tiny bit more cleanup.  Also can't land until the go assembly feature lands.  So marking draft again for now.

dkegel-fastly commented:
Rebased atop https://github.com/tinygo-org/tinygo/pull/2571 for the seek fix

dkegel-fastly commented:
https://github.com/tinygo-org/tinygo/pull/2571 should land first, but otherwise this is ready for review.

dkegel-fastly committed:
os: add DirFS, which is used by many programs to access readdir.

It's wafer-thin :-)

Includes smoke test from upstream.

TODO: once t.TempDir is implemented, add io/fs to the list of standard library tests to run; that's a better test.

dkegel-fastly commented on src/os/dir_darwin.go:
```diff
@@ -140,6 +140,12 @@ func darwinOpenDir(fd syscallFd) (uintptr, string, error) {
 			break
 		}
 	}
+	// Without this test, darwinOpenDir would happily open non-directories, causing trouble in at least Glob().
```
@aykevl syscall.Fdopendir needs reviewing, I think.

dkegel-fastly replied:
sorted.

dankegel commented:
I think this is ready, now that I've tested it with io/fs and fixed the problem this found on Darwin.

dkegel-fastly commented:
Oh, ok, maybe it's too late for 0.22.

dgryski reviewed (approved):
LGTM

deadprogram commented:
Thanks @dkegel-fastly and @dgryski now merging.

QuLogic commented on src/os/file_other.go:
```diff
@@ -22,6 +22,19 @@ const isOS = false
+	name   string
+}
+
+func NewFile(fd FileHandle, name string) *File {
```
This differs from Go in taking an interface for `fd` instead of a `uintptr`, which seems to break if passing, e.g., file descriptor 3 as done in [Go 1.18's fuzzing code](https://github.com/golang/go/blob/d588f487703e773ba4a2f0a04f2d4141610bff6b/src/internal/fuzz/sys_posix.go#L61) (cf. #2515)

dkegel-fastly replied:
IIRC that was taken straight from upstream?  Maybe upstream has changed, or maybe I'm misremembering.

Either way, please file a bug and we can sort it.

QuLogic replied:
Well, it's been that way for 10 years, it appears. I'll just open a PR as I've already made the changes locally.
